sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -509,18 +509,10 @@ abstract class StringPredicate extends BinaryExpression
       return checkResult
     }
     // Additional check needed for collation compatibility
-    val rightCollationId: Int = right.dataType.asInstanceOf[StringType].collationId
-    if (collationId != rightCollationId) {
-      DataTypeMismatch(
-        errorSubClass = "COLLATION_MISMATCH",
-        messageParameters = Map(
-          "collationNameLeft" -> CollationFactory.fetchCollation(collationId).collationName,
-          "collationNameRight" -> CollationFactory.fetchCollation(rightCollationId).collationName
-        )
-      )
-    } else {
-      TypeCheckResult.TypeCheckSuccess
-    }
+    val outputCollationId: Int = TypeCoercion
```

Comment:
why not add `StringPredicate` to the rule check like for `BinaryComparison`?

Reply:
Have looked into it. The problem we have is the test in `QueryCompilationErrorSuite`. We have 2 errors: `COLLATION_MISMATCH` and `UNSUPPORTED_COLLATION.FOR_FUNCTION`. If someone calls a function with different collations we run in to a problem which should be thrown first. Do we first try to collate parameters and then check if function is supported for that collation, or do we first throw an error that some collation is not supported by the function? If I leave the check in `StringPredicate`, the method `checkInputDataTypes` in super class is called first and we have `COLLATION_MISMATCH` first, but if I change it to the rule, we have `UNSUPPORTED_COLLATION.FOR_FUNCTION` first. The question is, which is better for the customer?

Reply:
I'd say `COLLATION_MISMATCH` first, `UNSUPPORTED_COLLATION.FOR_FUNCTION` second. If the user specifies COLLATION_1 for `left` and COLLATION_2 for `right`, how would we know which one to use when checking whether the functions supports this type of collation? (in this case, suppose a function supports COLLATION_1, but not COLLATION_2 - does the `UNSUPPORTED_COLLATION.FOR_FUNCTION` pass or fail?)

Hence, I think we would first need to establish that COLLATION_1 and COLLATION_2 are the same (no `COLLATION_MISMATCH `), before checking whether the function supports the requested collation (no `UNSUPPORTED_COLLATION.FOR_FUNCTION`)

Reply:
also note that in:
`final lazy val collationId: Int = left.dataType.asInstanceOf[StringType].collationId`

`left` is chosen arbitrarily, and shouldn't be used to check whether a function supports that collation type _before_ checking whether `right` has the same collation